### PR TITLE
Heal member-select on opaque reference

### LINKED ISSUE #19609: Opaque type not transparent in its defining scope, when referred to via qualified name
## Compiler version

3.3.1
3.3.2-RC3
3.4.0-RC4

## Minimized code

```Scala
object o {
  opaque type T = String

  summon[o.T =:= T]      // OK
  summon[o.T =:= String] // OK

  def test1(t: T): Int =
    t.length // OK

  def test2(t: o.T): Int =
    t.length // Error: value length is not a member of Playground.o.T
}
```

https://scastie.scala-lang.org/CN2ydzIkSBSKNUsY10NYgQ

## Output

```scala
value length is not a member of Playground.o.T
```

## Expectation

According to https://docs.scala-lang.org/scala3/reference/other-new-features/opaques-details.html#type-checking-10 

> we have inside the object (also for non-opaque types) that **o.T is equal to T or its expanded form o.this.T.** Equality is understood here as mutual subtyping, i.e. o.T <: o.this.T and o.this.T <: o.T. Furthermore, we have by the rules of opaque type aliases that **o.this.T equals R. The two equalities compose.** That is, inside o, it is also known that o.T is equal to R. 

it should hold, inside the object `o`, that `o.T` is equal to `o.this.T` is equal to `String`. Therefore, `t.length` should typecheck.

### PR DESCRIPTION
**Pull request by dwijnand**
When the prefix of an opaque isn't the .this reference of the module
class, then its RHS isn't visible.  TypeComparer uses ctx.owner to
"heal" or "lift" this type such that it is.  We reuse that logic for
member selection.

Fixes #19609

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 1]
```diff
@@ -688,72 +688,92 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
...
+        val qual1 = qual.cast(gadtApprox)
+        val tree1 = cpy.Select(tree0)(qual1, selName)
+        tryType(tree1, qual1, selectionType(tree1, qual1))
+          .orElse(trySmallGenericTuple(tree, qual1, withCast = false))
+          .orElse(tryExt(tree1, qual1))
+      else EmptyTree
+
+    tryType(tree, qual, rawType)
+      .orElse {
+        if selName == nme.apply && qual.tpe.widen.isInstanceOf[MethodType] then
```
Comment (sjrd) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:733]: Consider putting *all* the arguments to `.orElse`s in a separate `def`s? I'm not sure I understand the current mix, and it makes reading more difficult IMO.

Comment (sjrd) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala:733]: Also ideally this is the kind of things for which I like to have:
1. a refactoring commit that moves things around without changing the logic
2. the actual commit that changes the logic.

Here it takes a lot of brain power to identify what is actually changing.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 2]
```diff
@@ -688,72 +688,92 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
...
+      if checkedType.exists then
+        val select = toNotNullTermRef(assignType(tree, checkedType), pt)
+        if selName.isTypeName then checkStable(qual.tpe, qual.srcPos, "type prefix")
+        checkLegalValue(select, pt)
+        ConstFold(select)
+      else EmptyTree
+
+    def trySmallGenericTuple(tree: untpd.Select, qual: Tree, withCast: Boolean) =
+      if qual.tpe.isSmallGenericTuple then
+        if withCast then
```
Comment (sjrd) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:703]: Note to self: this `if withCast` seems to be new.

Edit: no, it's a boolean switch for the two call sites, which used to do the thing with and without the cast.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 3]
```diff
@@ -688,72 +688,92 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
...
+        if couldInstantiateTypeVar(qual.tpe.widen) then
+          // there's a simply visible type variable in the result; try again with a more defined qualifier type
+          // There's a second trial where we try to instantiate all type variables in `qual.tpe.widen`,
+          // but that is done only after we search for extension methods or conversions.
+          typedSelect(tree, pt, qual)
+        else EmptyTree
+      }
+      .orElse {
+        val wtp = qual.tpe.widen
+        val liftedTp = comparing(_.liftToThis(wtp))
```
Comment (sjrd) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:748]: Note to self: this is the actual new stuff.

**Review by sjrd — CHANGES_REQUESTED** [review]
My only real comment is to consider putting all the branches in separate `def`s. The rest is mostly rambling as I tried to understand the thing.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala (hunk 1 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala:733]: Fair on all points.  Will do.

**Review by sjrd — APPROVED** [review]

**Comment by som-snytt** [comment]
@dwijnand also fixes example 1 in https://github.com/scala/scala3/issues/15050

```
import OpaqueBug.*
def g(n: Counter): Counter = n
object OpaqueBug:
  opaque type Counter = Int
  val initial: Counter = 42
  def f(n: Int): Int = g(n) + initial

@main def run = println(OpaqueBug.f(21))
```

**Pull request merged**